packages/panzer/mini-em/src/eqn_sets/MiniEM_OperatorRequestCallback.cpp

```diff
@@ -50,8 +52,7 @@ Teko::LinearOp OperatorRequestCallback::request(const Teko::RequestMesg & rm)
 void OperatorRequestCallback::writeOut(const std::string & s,const Thyra::LinearOpBase<double> & op) const
 {
   using Teuchos::RCP;
-
-  typedef Tpetra::DefaultPlatform::DefaultPlatformType::NodeType NT;
+  using NT = panzer::TpetraNodeType;
```

Comment:
It looks like this was always the right thing to do anyway, since Panzer wants to derive the Node type from `PHX::Device`, rather than use Tpetra's default Node type.  (See `Panzer_NodeType.hpp`, which comments out `Tpetra::Map<>::node_type` and uses `PHX::Device` instead to derive Panzer's Node type.)  However, changes like this make me wonder if Panzer's Node type could ever have been different from Tpetra's default Node type.  I would much prefer a single source for defaults.  One of our desiderata is to get rid of Node type as a template parameter, in favor of having it be a configure-time choice; if Panzer needs the ability to make a nondefault choice, I'd like to know.  Thanks!

Reply:
@mhoemmen - I agree on the single choice for defaults, but am not sure how feasible that is. A year or so ago, we had one of these mismatched node issues. Phalanx only depends on kokkos. So it got its default from kokkos. Panzer depends on Tpetra and was using Tpetra default node. Both Tptera and Kokkos had same logic for picking defaults, but in an effort to save on compile time, a user disabled some Tpetra ETI node instantiations. So the kokkos default and tpetra default became mismatched. Unless we force Tpetra to compile all kokkos enabled node types for ETI, I can always see multiple defaults existing. I do not want to add an unneeded dependency on Tpetra to phalanx. Maybe Tpetra ETI has changed recently?

Management has made it clear that we should expect to enable and use multiple node types on future hardware and should design components accordingly. Simple example is to build kokkos with both OpenMP and CUDA enabled. Phalanx/Panzer runs on one node (not necessarily the default node) while another kokkos enabled application uses the other node in a coupled simulation. Along these lines, I hate the idea of having a default node type at all. Depending on how users implement the kokkos calls in library components, they could accidentally use an unintended node space. This is easy to do with UVM since memory movement is hidden. I would much rather have a slightly more verbose call stack where users specify node types in the application (still picked at compile time). This is why the panzer node type exists rather than always using the default node type. We force the users to explicitly chose where to run at configure time. Until we have a comprehensive plan for dealing with multiple node spaces, we force a single node space for all panzer kernels. chosen at configure time, that may not be the default node type. Not the ideal solution, but works for the time being...

Reply:
@rppawlo Here's the plan:

1. Tpetra's default Device WILL BE Kokkos' default Device (no choice) (this avoids the "user sets nondefault options that break the build" isssue mentioned in your first paragraph)
2. The Node template parameter will turn into a Kokkos execution space template parameter (improving ease of use and consistency with Kokkos)
3. Tpetra's methods will take an optional execution space instance, if you want to run somewhere else (this achieves the goal of the first sentence of your second paragraph)

My ultimate goal is that this "Node" or "Device" template parameter goes away.  Tpetra will use the default Kokkos execution space by default, but users will be able to supply a nondefault execution space instance if they want to run elsewhere.

Reply:
It's possible that users won't accept my "ultimate goal" -- e.g., they may want to enable CUDA in Kokkos but not in solvers -- but to me that says that solvers are doing the wrong thing.  If people want solvers to strong-scale, for example, it may make sense for Tpetra and solvers internally to switch over from GPU to CPU storage and execution, invisibly to the user.

Reply:
> It's possible that users won't accept my "ultimate goal" -- e.g., they may want to enable CUDA in Kokkos but not in solvers -- but to me that says that solvers are doing the wrong thing.

I don't think you'll have any issue convincing people of that. I could see a scenario where if one node type was restricted to explicit physics, but that's such a corner case. They'd probably still use Tpetra for ghosting.